Solution to first part

8n8 opened the pull request:
READY Hi, my solution to the first part of the challenge. It was fun, thank you.

8n8 committed:
Build script, instructions and empty test.

8n8 committed:
Added boilerplate to run the Elm code.

8n8 committed:
Basic exercise test for audiences decoder.

8n8 committed:
Fixed error in test: it was expecting Maybe but should have been expecting Result.

8n8 committed:
First draft of JSON decoder for audiences, but it is failing on type: user.

8n8 committed:
JSON decoder is passing simple tests.

8n8 committed:
Fixed issue with Audience decoder where it was expecting type: authored instead of type: user.

8n8 committed:
Fuzz test for Audience decoder.

8n8 committed:
Simple lsTest for folder decoder.

8n8 committed:
Folder decoder passes simple test.

8n8 committed:
Added encode-decode fuzz test for folder decoder.

8n8 committed:
Added empty app boilerplate.

8n8 committed:
Filled in 'init' function so that it starts out at the top of the tree.

8n8 committed:
First attempt passes compiler and tests, but clicking on folders doesn't show any sub-items.

8n8 committed:
Added failing test for child selector.

8n8 committed:
Added passing test for child selection helper.

8n8 committed:
Added failing test for getChildIds.

8n8 committed:
Fixed getChildIds. The view now looks much better, with stuff in the folders, but there is still an issue. When I click 'New Group' -> 'New Group 2', there is no content, and no 'Go up' button.

8n8 committed:
Improved Model so there is no longer any duplicate data, replacing current with parent. This fixes issue with empty folders behaving like root.

8n8 committed:
Changed styling so that folder buttons fit to their contents.

8n8 committed:
Shifted parsing into its own module. All the data is parsed and validated right at the start, so everything further on has good data.

8n8 committed:
Explained model design, and got rid of folders from model. audiences can be used instead.

8n8 committed:
Removed a load of junk.

8n8 committed:
Added some type declarations.

8n8 committed:
A few docs.

8n8 committed:
Added run instructions.

8n8 committed:
Fixed some dodgy indentation.

8n8 committed:
Stripped down Main exports.

8n8 committed:
Type, and removed unused datatype.

8n8 committed:
Tweak to comments, and removed empty module.

Janiczek commented:
Hey @8n8 thanks for the solution!
We'll take a look and review the code - but I just want to check - are you applying for the job or did you do the challenge just for fun? 🙂

8n8 commented:
Hi @Janiczek, I did it for fun mainly. Although I am looking for a part-time job, 2 days a week. I expect you want someone full time though.

Janiczek commented:
Correct. Ah, nevermind! :)

8n8 commented:
Hi @Janiczek, are you still hiring? I'm now able to work full time.

Janiczek commented on shell.nix:
```diff
@@ -0,0 +1,11 @@
+{ pkgs ? import (fetchTarball "https://github.com/NixOS/nixpkgs/archive/a165aeceda9f9741d15bc2488425daeb06c0707e.tar.gz") {} }:
+
+pkgs.mkShell {
```
Nice to see somebody use Nix :)

Janiczek commented on src/Main.elm:
```diff
@@ -1,10 +1,209 @@
+import Html.Attributes as Hat
+import Html.Events as Hev
+import Json.Decode as Jd
+import Parser as P
```
Reading this code, it was surprising seeing a `Parser` import - I immediately thought it's `elm/parser` parser. I'd keep the `Decoder` naming for stuff dealing with JSON, to prevent misunderstandings/surprises :)

8n8 replied:
I agree it's confusing, and have changed it to DataParser. I like having 'Parser' in the name because I like the way this module catches as many errors in the data as possible up front, i.e. I like the 'Parse not validate' school of thought.

Janiczek commented on src/Main.elm:
```diff
@@ -1,10 +1,209 @@
+    -- is root, so there are no internal errors caused by failing map
+    -- lookups. The reason for using a dictionary rather than a list
+    -- is to enforce that a child can have only one parent.
+    { parents : Dict.Dict Int Int
```
Do you know of a way to signal the difference between the Ints in code, instead of in comments? What options do we have, and which one you'd like the most?

8n8 replied:
I considered having something like:

```
type ChildId
    = ChildId Int
```

and 

```
type ParentId
    = ParentId Int
```

Or I could have type aliases instead of wrappers. One problem with the wrappers is that they are not comparable types in Elm (I think), so can't be used directly as Dict keys. One way to deal with this would be to make a module containing a Dict Int Int structure, but only export getters and setters for ChildId and ParentId. Something like:

```
module Parent exposing (Parents, ChildId, ParentId, makeChild, makeParent, setParent, getParent)

type Parents =
    Parents_ (Dict Int Int)

type ChildId = ChildId_ Int

type ParentId = ParentId_ Int

setParent : ChildId -> ParentId -> Parents -> Parents
setParent = ...

getParent : ChildId -> Parents -> ParentId
getParent = ...
```

I'm torn on this one. I really like the simplicity of plain Dict Int Int. I think it's also important to bear in mind that indirection has a readability cost, because whenever you see 'ChildId' in the code you have to look up its definition, which is one more bit of cognitive effort. And if it's in a separate module that's even harder. On the other hand, when mapping over the dict, functions will have arguments like ... Int -> Int ... and how can the reader tell which is which? They have to look back to where the function is called. There are the variable names, but they are not checked by the type system.

In general I'm not that keen on type aliases, unless they are for really huge types like record types that are not practical to include in full. They don't improve type safety, and increase mental indirection for the reader.

I am perhaps over-reacting slightly to some Haskell code I have seen that has so many type aliases and newtypes that it's really hard to know what's going on, and what the actual data is underneath. Elm doesn't seem to suffer from this problem so much.

I also feel that in this case it would be hard for the code to be subtly wrong if somewhere the child and parent IDs got mixed up. I think it would be noticeable straight away.

When in doubt, I tend to go with less code is better.

Janiczek replied:
> One problem with the wrappers is that they are not comparable types in Elm (I think), so can't be used directly as Dict keys.

FYI there are packages like [`turboMaCk/any-dict`](https://package.elm-lang.org/packages/turboMaCk/any-dict/latest/) (more performant, needs you to provide a `toComparable` function) and [`pzp1997/assoc-list`](https://package.elm-lang.org/packages/pzp1997/assoc-list/latest/) (less performant, but nicer library API) with which you can have a `Dict customtype a`

Janiczek replied:
IMHO even the type aliases would be better (would convey more meaning) than `Dict Int Int` as it could mean anything.

I would go with the custom type wrappers though.

In our real app code we go even further and do it with phantom types:

```elm
type Id a
    = Id String

type AudienceIdTag
    = AudienceIdTag

type alias AudienceId =
    Id AudienceIdTag

type AudienceFolderIdTag
    = AudienceFolderIdTag

type alias AudienceFolderId =
    Id AudienceFolderIdTag
```
which allows us to have some common logic for `Dict (Id tag) ...` and `Set (Id tag)`.

8n8 replied:
This dict is now gone, following your comments below, but one nice thing about using plain ints is that if I choose the IDs carefully, Dict Int Int could be changed to Array Int. So the child IDs are the array positions and the parent IDs are the array contents. I would think that the array lookups would be way faster than the dict lookups, but it's hard to be sure without benchmarking.

Janiczek replied:
Well, we can always benchmark!
```elm
module ArrayDictAccess exposing (main)

import Array
import Benchmark exposing (..)
import Benchmark.Runner exposing (BenchmarkProgram, program)
import Dict


main : BenchmarkProgram
main =
    program suite


suite : Benchmark
suite =
    let
        sampleDict =
            List.range 0 10000
                |> List.map (\i -> ( i, 3 * i ))
                |> Dict.fromList

        sampleArray =
            List.range 0 10000
                |> List.map (\i -> 3 * i)
                |> Array.fromList
    in
    Benchmark.compare "Access"
        "Dict"
        (\_ -> Dict.get 5000 sampleDict)
        "Array"
        (\_ -> Array.get 5000 sampleArray)
```
<img width="584" alt="Screenshot 2021-09-15 at 12 01 57" src="https://user-images.githubusercontent.com/149425/133413675-fff167a0-3580-465a-86d3-2f6278ede4d7.png">

8n8 replied:
Wow that looks good! Is that BrianHicks/elm-benchmark you're using? It says it's not compatible with 0.19.

Janiczek replied:
@8n8 Correct, it just got moved to [`elm-explorations/benchmark`](https://package.elm-lang.org/packages/elm-explorations/benchmark/latest/)! :)

8n8 replied:
I've just refactored it again, and now all the data is in arrays of basic types, with the array positions acting as IDs. There's a slight sacrifice of type safety, because I couldn't think of a way to guarantee in the types that everything has a name. But all the lookups are fast array lookups instead of dict lookups. I also added a breadcrumb.

8n8 replied:
I just thought of a way to avoid the loss of type safety, by giving the failing name lookups the meaning that the failing ID is the root folder, which doesn't have a name.

8n8 replied:
So now the whole internal state is just:
```
type alias GoodModel =
    { names : Array.Array String
    , parents : Array.Array Int
    , firstAudienceId : Int
    , currentLevel : Int
    }
```

Janiczek commented on src/Main.elm:
```diff
@@ -1,10 +1,209 @@
+    -- ever be possible to have an ID without a name. The reason for
+    -- using a dictionary instead of a list is to enforce that an
+    -- item can only have one name.
+    , all : Dict.Dict Int String
```
You're mixing audience and audience folder IDs then, which ties into your comment

> It seems likely that the provided IDs are row IDS from the database, and that folders and audiences are kept in different tables, so the IDs may well conflict. A globally unique ID is needed here.

Just a thought: it might be possible to have two dicts, one for audiences and one for audience folders. Then we wouldn't need to care about ID colisions nor have to assign them new IDs.

8n8 replied:
Yes I like this idea. And I just realised that if I assume names are unique within a folder, I don't need audience IDs at all, because an audience is uniquely identified by the combination of its name and its parent ID. So now I only have folder IDs and can use the original ones. In this new Model, I always store folder names alongside the folder IDs, because they always have a name, and this prevents having to look up names, which could fail. I can also drop 135 lines of complicated parser code, so hopefully it won't seem so over-engineered.

Janiczek commented on src/Main.elm:
```diff
@@ -1,10 +1,209 @@
+    -- using a dictionary instead of a list is to enforce that an
+    -- item can only have one name.
+    , all : Dict.Dict Int String
+    , parent : Parent
```
This naming seems non-ideal to me: it is being used as "where I am currently in the tree". Perhaps something like `currentlyOpen` or `currentLevel` might signal that better?

Janiczek replied:
Also, a question: is there some alternative way how to model this navigation breadcrumb problem without needing to look up whether a folder has a parent folder?

8n8 replied:
OK agreed, have changed 'parent' to 'currentLevel'.

I actually feel pretty happy with doing dict lookups to get the breadcrumb. The only objection I can think of is a performance one. But say there are usually less than 10 levels deep of folders, and now I have split folder parents into their own dict, say that there are usually around 1000ish folders in the system. That's 10 dict lookups in fairly small dicts for each folder click to get a proper breadcrumb like A -> B -> C -> D. This seems all right to me.

Janiczek replied:
Yeah I think the Dict approach is OK given the challenge constraints (we don't need to show a breadcrumb trail etc.).
Another solution I could think of is a stack of the IDs, so you would have
```elm
path : List AudienceFolderId
```
where the current parent is at the start and as you go to the right you get grandparent etc... all the way to the root.
That way any manipulations are very quick (adding / removing items from left) and you can eg. easily get the breadcrumbs with List.map and perhaps a List.reverse.

Janiczek commented on src/Main.elm:
```diff
@@ -1,10 +1,209 @@
+            ]
+
+
+goUpView : GoodModel -> List (Html.Html Msg)
```
Since this is a list of 0 or 1 items, it could be a Maybe (`if ... then Nothing else Just (Html.button ...)`).

Also since Html has an identity element (value of type `Html msg` that does nothing), namely `Html.Extra.nothing` which is just `Html.text ""`, this function could just return `Html Msg`: `if ... then Html.nothing else Html.button ...`

8n8 replied:
Oh that's neat. I had the impression that empty text would still affect the DOM somehow, but I've just done a little test and I see that it doesn't.

Janiczek commented on src/Main.elm:
```diff
@@ -1,10 +1,209 @@
+    List.map oneFolderView <|
+        Dict.toList <|
+            Dict.filter (isFolder model.audiences) <|
+                Dict.filter (isChildOf model.parents model.parent) model.all
```
Just a personal code preference: using `<|` causes indentation on each new step. Using `|>` doesn't do that:
```elm
model.all
    |> Dict.filter (isChildOf model.parents model.parent)
    |> Dict.filter (isFolder model.audiences)
    |> List.map oneFolderView
```

8n8 replied:
Yours looks neater visually, but I find it harder to understand. I think it's all the years of using languages that have the arguments in parenthesis has made me feel that arguments must go on the right. Perhaps I should try to get used to it this way.

Janiczek replied:
I definitely didn't mean it as "you SHOULD do it this way"! :) it's just that I find it nicer and wanted to share that possibility in case you didn't know it.

8n8 replied:
The more I look at it the more I like it.

Janiczek commented on src/Main.elm:
```diff
@@ -1,10 +1,209 @@
+
+
+isChildOf : Dict.Dict Int Int -> Parent -> Int -> String -> Bool
+isChildOf parents parent potentialChild _ =
```
This function has an unused `String` argument just because it's used in `Dict.filter`. It would be cleaner to remove the extra argument and call Dict.filter with `(\id _ -> isChildOf model.parents model.parent id)` IMHO.

8n8 replied:
I think I actually prefer my way slightly because at least the hole has an explicit type so is arguably easier to read.

Janiczek commented on src/Main.elm:
```diff
@@ -1,10 +1,209 @@
+
+isAudience : Dict.Dict Int A.AudienceType -> Int -> String -> Bool
+isAudience audiences id _ =
+    not <| Dict.get id audiences == Nothing
```
```suggestion
    Dict.get id audiences /= Nothing
```

8n8 replied:
Yes this is better.

Janiczek commented on README.md:
```diff
@@ -1,3 +1,32 @@
+
+## UI
+
+I decided not to spend time making it pretty, so there aren't even folder counts, or folder names at the top of opened folders. But these things are there in the model, so would be easy to put in.
```
When first running the project, I kept looking at whether I was missing some CSS file. A little polish wouldn't hurt, but anyways beauty wasn't the point of the challenge :)

8n8 replied:
If I get time I'll have a go at making it look nice. I would also like a proper breadcrumb.

Janiczek commented on src/Parser.elm:
```diff
@@ -0,0 +1,374 @@
+    , unique : Int
+    , folderIds : Dict.Dict Int Int
+    , audienceIds : Dict.Dict Int Int
+    }
```
Seems to me like it would be cleaner to create the fields separately using some List.map or List.indexedMap or List.foldl passes, rather than try to keep track of all the fields in one `addAudience`-like function.

Or at least I have a hard time verifying there is or is not a bug somewhere in them! :) With the separate computations it would be IMHO more localised and easier to check.

8n8 replied:
Yes I agree. I have now changed it so that the fields are accumulated separately, and it's also much simpler now it doesn't calculate new IDs, following your suggestion above.

Janiczek commented on src/Parser.elm:
```diff
@@ -0,0 +1,374 @@
+addSubAudience : MakeModel -> A.Audience -> Int -> Data
+addSubAudience m a parentId =
+    let
+        newId =
+            m.unique
+
+        ( newParentId, newUnique ) =
+            case Dict.get parentId m.folderIds of
+                Nothing ->
+                    ( m.unique + 1, m.unique + 2 )
+
+                Just exists ->
+                    ( exists, m.unique + 1 )
```
Re the previous comment near `MakeModel`: this for example is something I'd love to _not_ have to write. It could be circumvented using something like
```elm
let
  audiencesCount = List.length audiences
  foldersCount = List.length audienceFolders
  newAudienceIds = List.range 0 (audiencesCount - 1)
  newAudienceFolderIds = List.range audiencesCount (audiencesCount + foldersCount - 1)
in
...
```
and then just zipping them with the audiences list so that you have both the ID and the audience it should be applied to, in the same lambda.

But again, I think a cleaner solution would be to _not_ give new IDs to the audiences and folders.

Janiczek commented on src/Parser.elm:
```diff
@@ -0,0 +1,374 @@
+        ]
+
+
+allParentsExist : Dict.Dict Int String -> List ( Int, Int ) -> Maybe String
+allParentsExist all parents =
+    case parents of
+        [] ->
+            Nothing
+
+        ( child, parent ) :: arents ->
+            case Dict.get parent all of
+                Nothing ->
+                    Just <|
+                        String.concat
+                            [ "parent with ID "
+                            , String.fromInt parent
+                            , " and child "
+                            , String.fromInt child
+                            , " does not exist"
+                            ]
+
+                Just _ ->
+                    allParentsExist all arents
```
There are nicer solutions possible than the explicit recursion:
* if you're OK with just a `Bool` output (no further data about what's missing, just that _something's_ missing) then `List.all (parentExist all) parents` would do just fine
* if you want all errors, you can `List.filterMap ... parents`

The current solution returns the first error but not all of them.

8n8 replied:
I agree that mine is a bit ugly, but what I quite like is that it quits the map on the first error, which is what I want. Though I suppose it doesn't matter finding all of them; it might even help. In my new refactoring, I made a badMap function, for occasions when you want to quit the map when something goes wrong:

```
badMap : (a -> Result b c) -> List a -> Result b (List c)              
badMap f as_ =                                                         
    badMapHelp f as_ []                                                

                                                                       
badMapHelp : (a -> Result b c) -> List a -> List c -> Result b (List c)
badMapHelp f as_ accum =                                               
    case as_ of
        [] ->                                                          
            Ok (List.reverse accum)                                    

        a :: s ->                                                      
            case f a of
                Err err ->                                             
                    Err err                                            
    
                Ok ok ->                                               
                    badMapHelp f s (ok :: accum) 
```

Janiczek commented on src/Parser.elm:
```diff
@@ -0,0 +1,374 @@
+                            ]
+
+
+maybeChain : List (Maybe a) -> Maybe a
+maybeChain maybes =
+    case maybes of
+        [] ->
+            Nothing
+
+        Nothing :: aybes ->
+            maybeChain aybes
+
+        (Just m) :: _ ->
+            Just m
```
For fun, a possible reimplementation:
```elm
maybeChain maybes =
  maybes
    |> List.filterMap identity
    |> Maybe.andThen List.head
    |> Maybe.map Just
    |> Maybe.withDefault Nothing
```

```elm
maybeChain maybes =
    List.foldl Maybe.Extra.orElse Nothing maybes
```

8n8 replied:
These are fun. This |> style is growing on me. I think I need to use it more.

Janiczek commented on tests/Tests.elm:
```diff
@@ -0,0 +1,124 @@
+                Jd.decodeString
+                    P.decodeFolders
+                    F.audienceFoldersJSON
+    , fuzz folderFuzz "fuzz encode-decode for folder" <|
```
Kudos for fuzz tests!! :tada:

8n8 replied:
Thanks. Fuzzing is great for parsers.

Janiczek reviewed (approved):
The solution works - that's the main thing 🙂 

Sometimes it was a bit overengineered and IMHO there were some decisions early on in the type modelling phase that later led you to way too complicated solutions, but you've dealt with it 👍 

I've left some questions prodding for alternative solutions - if you have the time and are in the mood for it, you can try to answer these and we can discuss them further.

Thanks for the tip for the `Data-Oriented Design` book!

----

> Hi @Janiczek, are you still hiring? I'm now able to work full time.

Apologies for not reviewing the code earlier, and thanks for the ping!

I'll let @juanramoncg continue with the hiring process. AFAIK you can expect some Google Meet calls, and you'll probably be expected to send us your CV. But I'll let Juan handle the rest 🙂 

Thanks for the finished challenge!

8n8 committed:
Changed name of Parser module to DataParser to avoid confusion with elm/parser.

8n8 committed:
Restructured main Model. I no longer make new IDs for all the nodes, and don't use audience IDs at all.

8n8 committed:
Added elm-review to find unused code.

8n8 committed:
Removed some dead code found by elm-review.

8n8 committed:
Tidied up a few little things found by elm-review.

8n8 committed:
Refactored function that optionally generated HTML so that it uses a zero element rather than an empty list.

8n8 committed:
Replaced the large recursive functions in the parser with smaller map-like functions.

8n8 committed:
Added some type aliases in the main model, to document the meaning of basic values.

8n8 committed:
Tiny fix: changing String to AudienceName.

juanramoncg commented:
Hi @Janiczek thanks for introduce me and thanks @8n8 for doing the challenge. We have an open Senior FrontEnd position in our squad. If you don't mind I'd like to arrange a call so that I can talk a bit more about this position. Could you send me an email with your CV and availability to juan@gwi.com and then we can continue from there?

8n8 committed:
Changed the model so that the data is kept in columns, in arrays, to take advantage of fast array lookup.

8n8 committed:
Added a breadcrumb.

8n8 committed:
Added a couple of comments about the type safety sacrifice when using the Array-based model.

8n8 committed:
Fixed issue of having to handle invalid name lookups with an 'internal error' item name.

Items in root are now in 'parents' with a parent ID of -1. So a failing name lookup now has a valid meaning, which is that the item is root.

8n8 committed:
Fix to docstring on GoodModel.

juanramoncg commented:
Hi @8n8 not sure you got my previous message. Pinging!

8n8 commented:
Hi @juanramoncg, thanks for ping. I did see your previous message. I emailed you my CV on 16th Sep. Perhaps it ended up in Spam? I'll send it again.
